cpp/src/arrow/compute/expression_test.cc

```diff
@@ -630,6 +630,12 @@ TEST(Expression, BindWithImplicitCasts) {
                       literal(std::make_shared<TimestampScalar>(0, TimeUnit::NANO))),
                   cmp(field_ref("ts_s"),
                       literal(std::make_shared<TimestampScalar>(0, TimeUnit::SECOND))));
+    // GH-37110
+    ExpectBindsTo(
+        cmp(field_ref("ts_s"),
+            literal(std::make_shared<TimestampScalar>(0, TimeUnit::NANO, "UTC"))),
+        cmp(field_ref("ts_s"),
+            literal(std::make_shared<TimestampScalar>(0, TimeUnit::SECOND, "UTC"))));
```

Comment:
Since we are at it and the original PR didn't add many tests, can we also test non-zero values? For example (untested):
```suggestion
            literal(std::make_shared<TimestampScalar>(0, TimeUnit::SECOND, "UTC"))));
    ExpectBindsTo(
        cmp(field_ref("ts_s"),
            literal(std::make_shared<TimestampScalar>(123000, TimeUnit::NANO, "UTC"))),
        cmp(field_ref("ts_s"),
            literal(std::make_shared<TimestampScalar>(123, TimeUnit::MICRO, "UTC"))));         
```

Reply:
Added, I also found that we need a UTC

```c++
field("ts_s_utc", timestamp(TimeUnit::SECOND, "UTC"))
```